src/main/java/com/amazonaws/encryptionsdk/keyrings/KmsKeyring.java

```diff
@@ -0,0 +1,171 @@
+
+        // If the input encryption materials do not contain a plaintext data key and a generator is defined onEncrypt
+        // MUST attempt to generate a new plaintext data key and encrypt that data key by calling KMS GenerateDataKey.
+        if (!encryptionMaterials.hasPlaintextDataKey()) {
+            generateDataKey(encryptionMaterials);
+        } else {
+            // If this keyring's generator is defined and was not used to generate a data key, OnEncrypt
+            // MUST also attempt to encrypt the plaintext data key using the CMK specified by the generator.
+            keyIdsToEncrypt.add(generatorKeyId);
+        }
+
+        // Given a plaintext data key in the encryption materials, OnEncrypt MUST attempt
+        // to encrypt the plaintext data key using each CMK specified in it's key IDs list.
+        for (String keyId : keyIdsToEncrypt) {
+            encryptDataKey(keyId, encryptionMaterials);
+        }
+    }
+
+    private void generateDataKey(final EncryptionMaterials encryptionMaterials) {
+        final GenerateDataKeyResult result = dataKeyEncryptionDao.generateDataKey(generatorKeyId,
+                encryptionMaterials.getAlgorithmSuite(), encryptionMaterials.getEncryptionContext());
+
+        encryptionMaterials.setPlaintextDataKey(result.getPlaintextDataKey(),
+                new KeyringTraceEntry(KMS_PROVIDER_ID, generatorKeyId, KeyringTraceFlag.GENERATED_DATA_KEY));
+        encryptionMaterials.addEncryptedDataKey(result.getEncryptedDataKey(),
+                new KeyringTraceEntry(KMS_PROVIDER_ID, generatorKeyId, KeyringTraceFlag.ENCRYPTED_DATA_KEY, KeyringTraceFlag.SIGNED_ENCRYPTION_CONTEXT));
+    }
+
+    private void encryptDataKey(final String keyId, final EncryptionMaterials encryptionMaterials) {
+        final EncryptedDataKey encryptedDataKey = dataKeyEncryptionDao.encryptDataKey(keyId,
+                encryptionMaterials.getPlaintextDataKey(), encryptionMaterials.getEncryptionContext());
+
+        encryptionMaterials.addEncryptedDataKey(encryptedDataKey,
+                new KeyringTraceEntry(KMS_PROVIDER_ID, keyId, KeyringTraceFlag.ENCRYPTED_DATA_KEY, KeyringTraceFlag.SIGNED_ENCRYPTION_CONTEXT));
+    }
+
+    @Override
+    public void onDecrypt(DecryptionMaterials decryptionMaterials, List<? extends EncryptedDataKey> encryptedDataKeys) {
+        requireNonNull(decryptionMaterials, "decryptionMaterials are required");
+        requireNonNull(encryptedDataKeys, "encryptedDataKeys are required");
+
+        if (decryptionMaterials.hasPlaintextDataKey() || encryptedDataKeys.isEmpty()) {
+            return;
+        }
+
+        if (!encryptedDataKeys.stream()
+                .filter(edk -> edk.getProviderId().equals(KMS_PROVIDER_ID))
+                .map(edk -> new String(edk.getProviderInformation(), PROVIDER_ENCODING))
+                .allMatch(KmsUtils::isArnWellFormed)) {
+            throw new MalformedArnException("encryptedDataKeys contains a malformed ARN");
```

Comment:
My thinking was that since decryption failures are ignored, as a user of this method I would like the method to fail fast if I am passing in bad inputs. Otherwise I would not understand why a particular key was not being used for decryption.

Reply:
We should be flexible in what we accept (when it doesn't conflict with security). What if a newer version of the ESDK is delivered which supports a new form of ARNs? Older versions should still be able to decrypt data associated with the old-style of understood ARNs.

If no data keys can be decrypted, then this error should be propagated up. That's useful.

Reply:
I disagree.  The question is, how did such a value get in the the EDK in the first place?
I guess the question is "what is bad input"? if someone were to create a new KMS Keyring, that is not quite right and sometimes puts the wrong thing in the ProviderID, is this an example of a bad input?

Reply:
@seebees which point are you disagreeing with?